src/server/clients_info/client_visibility.rs

```diff
@@ -0,0 +1,284 @@
+                added,
+                removed,
+            } => {
+                if list.remove(&entity).is_some() {
+                    added.remove(&entity);
+                    removed.remove(&entity);
+                }
+            }
+        }
+    }
+
+    /// Returns an iterator of entities that lost visibility at this tick.
+    pub(super) fn iter_lost_visibility(&self) -> impl Iterator<Item = Entity> + '_ {
+        match &self.filter {
+            VisibilityFilter::All { .. } => VisibilityLostIter::AllVisible,
+            VisibilityFilter::Blacklist { added, .. } => {
+                VisibilityLostIter::Lost(added.iter().copied())
+            }
+            VisibilityFilter::Whitelist { removed, .. } => {
+                VisibilityLostIter::Lost(removed.iter().copied())
+            }
+        }
+    }
+
+    /// Sets visibility for specific entity.
+    ///
+    /// Does nothing if visibility policy for the server plugin is set to [`VisibilityPolicy::All`].
+    pub fn set_visible(&mut self, entity: Entity, visibile: bool) {
+        match &mut self.filter {
+            VisibilityFilter::All { .. } => {
+                if visibile {
+                    debug!(
+                        "ignoring visibility enable due to {:?}",
+                        VisibilityPolicy::All
+                    );
+                } else {
+                    warn!(
+                        "ignoring visibility disable due to {:?}",
+                        VisibilityPolicy::All
+                    );
+                }
+            }
+            VisibilityFilter::Blacklist {
+                list,
+                added,
+                removed,
+            } => {
+                if !visibile {
+                    if list.insert(entity, false).is_none() {
+                        removed.remove(&entity);
+                        added.insert(entity);
+                    }
+                } else if let Some(just_removed) = list.get_mut(&entity) {
+                    *just_removed = true;
+                    removed.insert(entity);
+                    added.remove(&entity);
+                }
+            }
+            VisibilityFilter::Whitelist {
+                list,
+                added,
+                removed,
+            } => {
+                if visibile {
+                    if list.insert(entity, true).is_none() {
+                        removed.remove(&entity);
+                        added.insert(entity);
+                    }
+                } else if list.remove(&entity).is_some() {
+                    removed.insert(entity);
+                    added.remove(&entity);
+                }
+            }
+        }
+    }
+
+    /// Gets visibility for specific entity.
+    pub fn is_visible(&self, entity: Entity) -> bool {
+        match &self.filter {
+            VisibilityFilter::All { .. } => true,
+            VisibilityFilter::Blacklist { list, .. } => !list.contains_key(&entity),
```

Comment:
> You don't remove the entity from the blacklist list in set_visible(entity, true), so this is incorrect.

This is expected. Since I want to minimize lookup, I store a boolean which indicates that the entity was queued for removal in this tick. Then later `cache_entity_visibility` will do a single lookup into `list` without additionally checking `removed`.
I probably should write a comment about it. It's really confusing.

I haven’t tested despawns, I just forgot, but I’ll do it in this PR. I wrote about this before in Discord.

Reply:
Here is where I check it:
https://github.com/lifescapegame/bevy_replicon/blob/e6bac4b23b9628af602cbe9ec179f18c44c271a1/src/server/clients_info/client_visibility.rs?plain=1#L206-L210